Added all possible conditions to http_cors module

aman566 opened the pull request:
CORS misconfiguration is not only about seeing Access-Control-Allow-Origin: * response header but also Access-Control-Allow-Origin: http://example.foo\r\nAccess-Control-Allow-Credentials: True. If these headers are present that for sure the website is vulnerable to cors.

#### Checklist
- [X] I have followed the [Contributor Guidelines](https://github.com/zdresearch/OWASP-Nettacker/wiki/Developers#contribution-guidelines).
- [X] I have added the relevant documentation.
- [X] My branch is up-to-date with the Upstream master branch.

#### Changes proposed in this pull request

#### Your development environment
- OS: `Linux`
- OS Version: `Ubuntu`
- Python Version: `3.6.9`

the-Blue-Falcon commented on lib/vuln/http_cors/engine.py:
```diff
@@ -75,6 +75,8 @@ def http_cors(target, port, timeout_sec, log_in_file, language, time_sleep,
             req = requests.get(target, headers=headers)
             if req.headers['Access-Control-Allow-Origin'] == "*":
                 return True
+            elif req.headers['Access-Control-Allow-Origin'] == "http://example.foo" and req.headers['Access-Control-Allow-Credentials'] == "true":
```
Why in the actual reality would you check for `http://example.foo`? Checking for CORS misconfiguration involves checking for the following:
- wildcard value (implemented)
- third party allowance (not only `http://example.foo` lol)
- origin reflection
- invalid value in ACAO header
- post-domain wildcard
- pre-domain wildcard
- null-origin allowance
- http-origin allowance
- unescaped regular expressions

aman566 replied:
Thank you for the suggestion. I added all the factors.

aman566 commented:
@securestep9 Please review it and suggest me some changes if any :)

Ali-Razmjoo commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Clones removed
==============
+ lib/vuln/http_cors/engine.py  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zdresearch/OWASP-Nettacker/pullRequest?prid=5173581&bid=16983066)

Ali-Razmjoo commented on lib/vuln/http_cors/engine.py:
```diff
@@ -23,6 +23,9 @@
 from core._time import now
 from core.log import __log_into_file
 import requests
+from urlparse import urlparse, parse_qs
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused parse_qs imported from urlparse](https://app.codacy.com/manual/zdresearch/OWASP-Nettacker/pullRequest?prid=5173581)

Ali-Razmjoo commented on lib/vuln/http_cors/engine.py:
```diff
@@ -23,6 +23,9 @@
 import requests
+from urlparse import urlparse, parse_qs
+import tldextract
+import time
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Reimport 'time' (imported line 7)](https://app.codacy.com/manual/zdresearch/OWASP-Nettacker/pullRequest?prid=5173581)

Ali-Razmjoo commented on lib/vuln/http_cors/engine.py:
```diff
@@ -70,10 +73,75 @@ def http_cors(target, port, timeout_sec, log_in_file, language, time_sleep,
+            domainName =  tldextract.extract(target)
+            root = domainName.domain + domainName.suffix
+            origin = {
+                'wildcard value': '*', 
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Trailing whitespace](https://app.codacy.com/manual/zdresearch/OWASP-Nettacker/pullRequest?prid=5173581)

aman566 commented:
@Ali-Razmjoo @the-Blue-Falcon Please review it. I have added everything what @the-Blue-Falcon said. Sorry @the-Blue-Falcon for not adding it. I have added it now. Please review it :)

itsdivyanshjain commented:
@securestep9 
I think it is important to have all the conditions in https_cors module. But @aman566 can you use nettacker logging function instead of using print function.

aman566 commented:
Hello @itsdivyanshjain can you please mention which all cases I am missing.
About the logging function, I will change it for sure :)

itsdivyanshjain commented:
@aman566 i am referring to @securestep9 that currently master branch has very few condition for cors and this PR is important to merge because you mentioned mostly all the condition.

But still I have noticed now that you have not mentioned some condition like unrecognized underscore, thirdparty allowed.

you can refer this https://github.com/s0md3v/Corsy for all the conditions.

Ali-Razmjoo commented:
this can't be supported anymore. modules are yaml #440
